src/instrumentation/create-telemetry-middleware.ts

```diff
@@ -0,0 +1,321 @@
+      if (lastIterationOnNewTokenSpanId && meta.name === newTokenEventName) {
+        // delete span
+        cleanSpanSources({ spanId: lastIterationOnNewTokenSpanId });
+        spansMap.delete(lastIterationOnNewTokenSpanId);
+      }
+
+      // delete the last `partialUpdate` event if the new one has same data and the original one does not have nested spans
+      const lastIterationEventSpanId = eventsIterationsMap.get(lastIteration)?.get(meta.name);
+      if (
+        lastIterationEventSpanId &&
+        ([partialUpdateEventName] as string[]).includes(meta.name) &&
+        spansMap.has(lastIterationEventSpanId)
+      ) {
+        const { attributes, context } = spansMap.get(lastIterationEventSpanId)!;
+
+        if (isDeepEqual(serializedData, attributes.data)) {
+          if (parentIdsMap.has(context.span_id)) {
+            spansToDeleteMap.set(lastIterationEventSpanId, undefined);
+          } else {
+            // delete span
+            cleanSpanSources({ spanId: lastIterationEventSpanId });
+            spansMap.delete(lastIterationEventSpanId);
+          }
+        }
+      }
+
+      // create new span
+      spansMap.set(span.context.span_id, span);
+      // update number of nested spans for parent_id if exists
+      if (span.parent_id) {
+        parentIdsMap.set(span.parent_id, (parentIdsMap.get(span.parent_id) || 0) + 1);
+      }
+
+      // save the last event for each iteration
+      if (groupIterations.length > 0) {
+        if (eventsIterationsMap.has(lastIteration)) {
+          eventsIterationsMap.get(lastIteration)!.set(meta.name, span.context.span_id);
+        } else {
+          eventsIterationsMap.set(lastIteration, new Map().set(meta.name, span.context.span_id));
+        }
+      }
+    });
+
+    // The generated response and message history are collected from the `success` event
+    emitter.on(
+      successEventName as any,
+      (data: InferCallbackValue<BeeCallbacks[typeof successEventName]>) => {
+        if ("data" in data && "memory" in data) {
+          const { data: dataObject, memory } = data as InferCallbackValue<
+            BeeCallbacks[typeof successEventName]
+          >;
+
+          generatedMessage = {
+            role: dataObject.role,
+            text: dataObject.text,
+          };
+          history = memory.messages.map((msg) => ({ text: msg.text, role: msg.role }));
+        }
+      },
+    );
+
+    // Read rawPrompt from llm input only for supported adapters and create the custom event with it
+    emitter.match(
+      (event) => "messagesToPrompt" in event.creator && event.name === startEventName,
+      ({ input }: InferCallbackValue<GenerateCallbacks[typeof startEventName]>, meta) => {
+        if (
+          "messagesToPrompt" in meta.creator &&
+          typeof meta.creator.messagesToPrompt === "function" &&
+          meta.creator instanceof Serializable &&
+          meta.trace
+        ) {
```

Comment:
Not sure if I even want to do it instead of just collecting messages.

Reply:
yes, we need to collect some important information from the framework like **rawPrompt**:
```
const rawPrompt = meta.creator.messagesToPrompt(input);
```
if is there any other way, how to collect this information, let me know.